### ISSUE TITLE
MAINT: `stats.dirichlet_multinomial`: relax `n` to `>=0`

### ISSUE THREAD

**Issue by j-towns** [scipy/scipy#21531]
The methods in `scipy.stats.dirichlet_multinomial` (which was added [relatively recently](https://github.com/scipy/scipy/pull/17211)) call the function [`_dirichlet_multinomial_check_parameters`](https://github.com/scipy/scipy/blob/92d2a8592782ee19a1161d0bf3fc2241ba78bb63/scipy/stats/_multivariate.py#L5988-L6018) to validate their arguments.

There is a requirement there that `n > 0`, which is stricter than necessary. The pmf, logpmf, mean, covariance and variance all have well defined values in the case where zero trials have taken place, so `n >= 0` should be sufficient. Moreover, `scipy.stats.multinomial` [only rejects `n < 0`](https://github.com/scipy/scipy/blob/92d2a8592782ee19a1161d0bf3fc2241ba78bb63/scipy/stats/_multivariate.py#L3277), and the dirichlet-multinomial behaviour should be consistent with that of the multinomial.

@MatteoRaso was there a reason for the positivity requirement? I'm happy to write a pr for this if no-one disagrees with me.

**Comment by mdhaber** [comment]
Thanks for your interest in improving SciPy. I don't remember the specific reason for the positivity requirement, but it may have been that it is easier to disallow it and raise an informative error than to check all possible edge case behavior. To better determine how time should be allocated, can you explain the need to use this distribution with `n = 0`?

**Comment by j-towns** [comment]
Thanks for getting back to me.

This is my use case: a Dirichlet prior is used to generate each row of the transition matrix of a Markov chain. I run the Markov chain and want to get the marginal log probability of the runs, which can be done by applying the Dirichlet-multinomial logpmf to the rows of the matrix of transition counts. States which never occurred have a row of zeros in this matrix, with `n  == 0`. Working around the fact that these zeros aren't allowed slows my code down.

I can type out a standalone version of the code that is broken and workaround if it's helpful.

It looks like simply changing `n <= 0` to `n < 0` on [this line](https://github.com/scipy/scipy/blob/92d2a8592782ee19a1161d0bf3fc2241ba78bb63/scipy/stats/_multivariate.py#L6011) gives the results I'd expect:
```python
In [1]: from scipy.stats import dirichlet_multinomial
In [2]: import numpy as np
In [3]: dirichlet_multinomial.mean(np.ones((3,)), 0)
Out[3]: array([0., 0., 0.])
In [4]: dirichlet_multinomial.cov(np.ones((3,)), 0)
Out[4]: 
array([[0., 0., 0.],
       [0., 0., 0.],
       [0., 0., 0.]])
In [5]: dirichlet_multinomial.var(np.ones((3,)), 0)
Out[5]: array([0., 0., 0.])
In [6]: dirichlet_multinomial.pmf(np.zeros((3,)), np.ones((3,)), 0)
Out[6]: np.float64(1.0)
In [7]: dirichlet_multinomial.logpmf(np.zeros((3,)), np.ones((3,)), 0)
Out[7]: np.float64(0.0)
```
I understand we need to be careful though and make sure there aren't any corner cases that break when we do this.

**Comment by j-towns** [comment]
Thanks for the info @mdhaber, that makes a lot of sense. I spent a little time looking for academic articles and books which explicitly specify the valid range of `n`. I couldn't find any, so (a few weeks ago) I edited the Wikipedia article, hoping that if the change there is uncontroversial then you would accept that as a reference.

I've also submitted a [pr](https://github.com/scipy/scipy/pull/21684).

**Comment by mdhaber** [comment]
I am uncomfortable with that precedent. If someone else had independently written that the domain is `n >= 0`, then I would know that at least 100% more people think that is appropriate. 

I am not motivated to push for SciPy to tread new ground here, but I won't get in the way. I will leave the issue and PR open for a few months to see if others are interested.

**Comment by j-towns** [comment]
OK, thanks. FYI (for anyone reading this in future) I also edited the Wikipedia [multinomial distribution](https://en.wikipedia.org/wiki/Multinomial_distribution) article to `n >= 0`. SciPy's multinomial is already `n >= 0`, as mentioned above.

**Comment by fancidev** [comment]
Although I don’t use this distribution, I feel it’s ok to allow the number of trials to be zero, as it extends the domain without going beyond the defining properties of the distribution, and without causing numerical difficulty.

This does require to review the formulas more carefully, but once reviewed I think it’s actually a good thing because the fact that the formulas work for “corner cases” suggests to me that they are likely well defined.

This “extension” reminds me of the definition of `all([])` and `any([])` — they’re examples of well defined “corner cases” that play well with e.g. recursion identities.

**Comment by j-towns** [comment]
>This “extension” reminds me of the definition of all([]) and any([]) — they’re examples of well defined “corner cases” that play well with general usage (e.g. those involving recursion identities).

Yes, also `sum([])`. It's convenient to use a list of length zero, and it's also convenient to compute the statistics of a distribution which is the sum of zero independent trials.
